crypto/hpke/hpke.c

```diff
@@ -0,0 +1,1537 @@
+/*
+ * Copyright 2022 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+/* An OpenSSL-based HPKE implementation of RFC9180 */
+
+#include <string.h>
+#include <openssl/rand.h>
+#include <openssl/kdf.h>
+#include <openssl/core_names.h>
+#include <openssl/hpke.h>
+#include <openssl/sha.h>
+#include <openssl/evp.h>
+#include <openssl/err.h>
+#include "internal/hpke_util.h"
+#include "internal/nelem.h"
+
+/** default buffer size for keys and internal buffers we use */
+#ifndef OSSL_HPKE_MAXSIZE
+# define OSSL_HPKE_MAXSIZE 512
+#endif
+
+/* Define HPKE labels from RFC 9180 in hex for EBCDIC compatibility */
+/**< "HPKE" - "suite_id" label for section 5.1 */
+static const char OSSL_HPKE_SEC51LABEL[] = "\x48\x50\x4b\x45";
+/**< "psk_id_hash" - in key_schedule_context */
+static const char OSSL_HPKE_PSKIDHASH_LABEL[] = "\x70\x73\x6b\x5f\x69\x64\x5f\x68\x61\x73\x68";
+/**<  "info_hash" - in key_schedule_context */
+static const char OSSL_HPKE_INFOHASH_LABEL[] = "\x69\x6e\x66\x6f\x5f\x68\x61\x73\x68";
+/**<  "base_nonce" - base nonce calc label */
+static const char OSSL_HPKE_NONCE_LABEL[] = "\x62\x61\x73\x65\x5f\x6e\x6f\x6e\x63\x65";
+/**<  "exp" - internal exporter secret generation label */
+static const char OSSL_HPKE_EXP_LABEL[] = "\x65\x78\x70";
+/**<  "sec" - external label for exporting secret */
+static const char OSSL_HPKE_EXP_SEC_LABEL[] = "\x73\x65\x63";
+/**<  "key" - label for use when generating key from shared secret */
+static const char OSSL_HPKE_KEY_LABEL[] = "\x6b\x65\x79";
+/**<  "psk_hash" - for hashing PSK */
+static const char OSSL_HPKE_PSK_HASH_LABEL[] = "\x70\x73\x6b\x5f\x68\x61\x73\x68";
+/**<  "secret" - for generating shared secret */
+static const char OSSL_HPKE_SECRET_LABEL[] = "\x73\x65\x63\x72\x65\x74";
+
+/* "strength" input to RAND_bytes_ex */
+#define OSSL_HPKE_RSTRENGTH 10
+
+/**
+ * @brief sender or receiver context
+ */
+struct ossl_hpke_ctx_st
+{
+    OSSL_LIB_CTX *libctx; /**< library context */
```

Comment:
What does the '*<' notation mean? (Just looks like a typo)..

Reply:
> Here is a NIST doc that treats the types differently.. https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.186-5-draft.pdf It refers to ECC curves and Edwards Curves. (Search for X25519 and P256). Anything listed in this doc is a NIST curve.. i.e They are approved for FIPS mode.

So yes, internally different curves will have different treatments but those aren't (and IMO oughtn't) be visible from "outside" which in this case means users of the HPKE API. Within the API implementation we need to deal with the fact that OpenSSL currently has different APIs for these two sets of curves (NIST/CFRG or ec/ecx). So we agree there's a need for a ``this_way_or_that_way()`` internal function, we're just disgreeing on the name.

The name you suggests isn't good IMO as "is_ec_curve" is true for CFRG curves (and all curves), the fact that "ec" has been used within OpenSSL to mean the curves defined by NIST for which there are HPKE codepoints may imply using that name, but nonetheless it'd be quite a surprise (at first) if ``is_ec_curve(curve25519)`` were to return 0. 
It is also true that NIST first standardised the p256, p384 and p521 curves, that those are commonly referred to as NIST curves, and that x25519 and x448 are similarly referred to as CFRG curves.

I'm not sure how OpenSSL handles brainpool curves though - perhaps that'd help distinguish?
Or, we could reverse the logic and define an ``is_cfrg_curve()`` function - that'd work for me if it helped.

Reply:
A fairly quick check didn't seem to show up anything useful in terms of getting value from doxygen when the external APIs and types are documented in a pod file, so I'll scrub those doxygen annotations from the next commit.